sycl/doc/extensions/SubGroupBitmaskGroupFunctions/sub_group_bitmask_group_algorithms.asciidoc

```diff
@@ -0,0 +1,253 @@
+
+`ext_oneapi_partition_sub_group` will create a bitmask with contiguous bits set to represent a partition of a subgroup.
+The size of the partition must be a power of two and less than or equal to  `sub_group_mask::max_bits`.
+The partition represented will be the partition the calling work-item is part of.
+For example, if `sub_group::get_local_linear_range()` is 32, `sub_group::get_local_linear_id()` is 18,
+and the partition size is 8, then the bitmask created would be `0b00000000 00000000 11111111 00000000`.
+Partitioning sub-groups to compile-time known sizes can allow for optimisations like loop unrolling.
+Care should be taken when modifying the value of the returned `sub_group_mask`
+since this will change the behavior of masked group functions.
+
+`ext_oneapi_active_sub_group_items` will create a bitmask with bits set to represent the
+work-items that are actively executing the command at the same time. Querying
+the currently active work-items should not be used to select the work-items that
+follow the same control flow since no synchronisation takes place; a
+`group_ballot` call will synchronize the work-items before testing the predicate, so is
+likely the best option. Querying the active work-items can be helpful for
+amortizing the cost of operations where unknown size groups of work-items
+could opportunistically perform the action together, such as atomic operation.
+Some modern architecture, like NVIDIA's Volta and later, do not guarantee
+synchronized forward progress of work-items in the same sub-group. This
+divergence is why work-items following the same control flow may not be
+active at the same time. Care should be taken when modifying the value of the 
+returned `sub_group_mask` since this will change the behavior of masked group functions.
+
+`ext_oneapi_rank_in_mask` returns an integer id representing the calling work-item's
+position within sub-group subset represented by the provided `sub_group_mask`. This is
+undefined behavior if bit `n` of the `sub_group_mask` is not set, where `n` is the
+integer id of the work-item in the sub-group. As an example, if the bitmask was 
+`0b00000000 00000000 11110000 11110000` and `get_sub_group().get_local_linear_id() == 13`,
+ext_oneapi_rank_in_mask would return 5.
+
+
+```c++
+namespace sycl {
+template <int dimensions = 1>
+class nd_item {
+  public:
+...
+  // create a bitmask of the sub-group subset this work-item would be a part of
+  // if the sub-group was split into contiguous groups of size `partition_size`
+  sub_group_mask ext_oneapi_partition_sub_group(size_t partition_size) const;
+
+  // get the bitmask representing the work-items of the sub-group that are
+  // actively executing.
+  sub_group_mask ext_oneapi_active_sub_group_items() const;
+
+  //Returns an integer id representing the calling work-item's position within the provided sub_group_mask.
+  size_t ext_oneapi_rank_in_mask(sub_group_mask mask) const;
```

Comment:
Bikeshedding: This form of "rank" doesn't appear in the SYCL 2020 specification.  I can see that using "id" here instead would be awkward, but if we could come up with a form that used "id" somehow I think it would be more consistent.

Reply:
Alright, that makes sense. I think we could go for `get_mask_id` and then hopefully the name is odd enough that a user could look up the definition the first time.
If we went for a proposal that used classes, like #4904, it could be just `get_group_id` which would fit nicely.